gen_stub.php

```diff
@@ -0,0 +1,1998 @@
+                    continue;
+                }
+
+                if (!$classStmt instanceof Stmt\ClassMethod) {
+                    throw new Exception("Not implemented {$classStmt->getType()}");
+                }
+
+                $classFlags = 0;
+                if ($stmt instanceof Class_) {
+                    $classFlags = $stmt->flags;
+                }
+
+                $flags = $classStmt->flags;
+                if ($stmt instanceof Stmt\Interface_) {
+                    $flags |= Class_::MODIFIER_ABSTRACT;
+                }
+
+                if (!($flags & Class_::VISIBILITY_MODIFIER_MASK)) {
+                    throw new Exception("Method visibility modifier is required");
+                }
+
+                $methodInfos[] = parseFunctionLike(
+                    $prettyPrinter,
+                    new MethodName($className, $classStmt->name->toString()),
+                    $classFlags,
+                    $flags,
+                    $classStmt,
+                    $cond
+                );
+            }
+
+            $fileInfo->classInfos[] = new ClassInfo($className, $methodInfos);
+            continue;
+        }
+
+        throw new Exception("Unexpected node {$stmt->getType()}");
+    }
+}
+
+function parseStubFile(string $code): FileInfo {
+    $lexer = new PhpParser\Lexer();
+    $parser = new PhpParser\Parser\Php7($lexer);
+    $nodeTraverser = new PhpParser\NodeTraverser;
+    $nodeTraverser->addVisitor(new PhpParser\NodeVisitor\NameResolver);
+    $prettyPrinter = new class extends Standard {
+        protected function pName_FullyQualified(Name\FullyQualified $node) {
+            return implode('\\', $node->parts);
+        }
+    };
+
+    $stmts = $parser->parse($code);
+    $nodeTraverser->traverse($stmts);
+
+    $fileInfo = new FileInfo;
+    $fileDocComment = getFileDocComment($stmts);
+    if ($fileDocComment) {
+        $fileTags = parseDocComment($fileDocComment);
+        foreach ($fileTags as $tag) {
+            if ($tag->name === 'generate-function-entries') {
+                $fileInfo->generateFunctionEntries = true;
+                $fileInfo->declarationPrefix = $tag->value ? $tag->value . " " : "";
+            } else if ($tag->name === 'generate-legacy-arginfo') {
+                $fileInfo->generateLegacyArginfo = true;
+            }
+        }
+    }
+
+    handleStatements($fileInfo, $stmts, $prettyPrinter);
+    return $fileInfo;
+}
+
+function funcInfoToCode(FuncInfo $funcInfo, bool $minimal): string {
+    $code = '';
+
+    // Generate the minimal, most compatible arginfo across PHP versions
+    if ($minimal) {
+        $code .= sprintf("ZEND_BEGIN_ARG_INFO_EX(%s, 0, %d, %d)\n",
+            $funcInfo->getArgInfoName(),
+            $funcInfo->return->byRef,
+            $funcInfo->numRequiredArgs);
+        foreach ($funcInfo->args as $argInfo) {
+            $code .= sprintf("\tZEND_ARG_INFO(0, %s)\n", $argInfo->name);
+        }
+    } else {
```

Comment:
This is the major change I made, the rest is just imported

Reply:
I took this technique from the bz2 extension, it's the least information you can provide and still have arginfo so you can use the partially validated extension parameter API. They haven't modified the arginfo in that extension since 5.5, and we don't use the new (7.2+) macro-based parameter API, so arginfo is basically useless to us.